clang/runtime/dpct-rt/include/dpct/bindless_images.hpp

```diff
@@ -117,6 +151,349 @@ class image_mem_wrapper {
+      ext_mem_win_handle{_win_nt_handle};
+
+    sycl::ext::oneapi::experimental::external_mem_descriptor
+      <sycl::ext::oneapi::experimental::resource_win32_handle> ext_mem_desc{
+        ext_mem_win_handle,
+        sycl::ext::oneapi::experimental::external_mem_handle_type::
+            win32_nt_dx12_resource,
+        _res_size_bytes};
+
+    _interop_mem_handle_ptr =
+        new sycl::ext::oneapi::experimental::interop_mem_handle;
+    *_interop_mem_handle_ptr =
+          sycl::ext::oneapi::experimental::import_external_memory(
+              ext_mem_desc, get_default_queue());
+
+    uniqueD3D11Res.insert(_d3d11_res);
+  }
+
+  ~interop_mem_wrapper() {
+    if (!_interop_mem_handle_ptr) {
+      throw std::runtime_error("Resource is not registered!");
+    }
+
+    CloseHandle(_win_nt_handle);
+
+    sycl::ext::oneapi::experimental::release_external_memory(
+        get_interop_mem_handle(), get_default_queue());
+    
+    uniqueD3D11Res.erase(_d3d11_res);
+  }
+#endif
+
+  interop_mem_wrapper(const interop_mem_wrapper &) = delete;
+
+  interop_mem_wrapper &operator=(const interop_mem_wrapper &) = delete;
+
+  const sycl::ext::oneapi::experimental::interop_mem_handle
+  get_interop_mem_handle() const noexcept {
+    return *_interop_mem_handle_ptr;
+  }
+
+  sycl::ext::oneapi::experimental::image_mem_handle
+  get_img_mem_handle() const noexcept {
+    return _img_mem_wrapper_ptr->get_handle();
+  }
+
+  void *
+  get_buf_ptr() const noexcept {
+    return _buf_ptr;
+  }
+
+  sycl::ext::oneapi::experimental::image_descriptor
+  get_img_desc() const noexcept {
+    return _img_desc;
+  }
+
+  size_t get_res_size_bytes() {
+    return _res_size_bytes;
+  }
+
+  bool is_buffer_res() const {
+    return _res_is_buffer;
+  }
+
+  void map_resource(sycl::queue q) {
+    if (_res_is_buffer) {
+      if (_buf_ptr) {
+        throw std::runtime_error("Resource is already mapped!");
+      }
+
+      // NOTE: 'map_external_linear_memory' API is not yet merged by Codeplay
+      /*
+      _buf_ptr =
+          sycl::ext::oneapi::experimental::map_external_linear_memory(
+              get_interop_mem_handle(), _img_desc, q);
+      */
+      throw
+        std::runtime_error(
+            "Mapping a buffer resoure is not yet supported!");
+    }
+    else {
+      if (_img_mem_wrapper_ptr) {
+        throw std::runtime_error("Resource is already mapped!");
+      }
+
+      auto img_mem_handle =
+          sycl::ext::oneapi::experimental::map_external_image_memory(
+              get_interop_mem_handle(), _img_desc, q);
+      
+      _img_mem_wrapper_ptr =
+          new image_mem_wrapper(get_img_desc(), img_mem_handle);
+    }
+  }
+
+  image_mem_wrapper* get_mapped_mipmapped_array() {
+    if(_res_is_buffer) {
+      throw std::runtime_error(
+          "Buffer resouce cannot be accessed as an array!");
+    }
+    if(!_img_mem_wrapper_ptr) {
+      throw std::runtime_error("Resouce is not mapped!");
+    }
```

Comment:
Should it throw an exception if `_img_mem_wrappter_ptr` is not a mipmapped array?

Reply:
There is no mention of this check in the CUDA [doc](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__INTEROP.html#group__CUDART__INTEROP_1g84c3772d2ed06cda8c92bc43cdc893d0:~:text=If%20resource%20is%20not%20a%20texture%20then%20it%20cannot%20be%20accessed%20via%20an%20array%20and%20cudaErrorUnknown%20is%20returned.%20If%20resource%20is%20not%20mapped%20then%20cudaErrorUnknown%20is%20returned.) for `cudaGraphicsResourceGetMappedMipmappedArray `.

Besides, there are only 2 ways to retrieve the underlying image data of resource in CUDA - 
- `cudaGraphicsResourceGetMappedMipmappedArray(array, resource)`
- `cudaGraphicsSubResourceGetMappedArray(array, resource, arrayIndex, mipLevel)`

As any non-buffer (texture) resource qualifies as a mipmap image with 1 level, I believe that's the reason why the image returned by `cudaGraphicsResourceGetMappedMipmappedArray` is not checked for having multiple levels

Instead, this check is done for `cudaGraphicsSubResourceGetMappedArray` ([ref](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__INTEROP.html#group__CUDART__INTEROP_1g84c3772d2ed06cda8c92bc43cdc893d0:~:text=If%20mipLevel%20is%20not%20a%20valid%20mipmap%20level%20for%20resource%20then%20cudaErrorInvalidValue%20is%20returned.)), which is used to get a specific level of a mipmap image.

Please correct/suggest my thoughts on this.